### PR TITLE
eip165 - Pseudo-Introspection, or standard interface detection

### PR DESCRIPTION
**Pull request by jbaylina**
Formal PR for Pseudo-Introspection of https://github.com/ethereum/EIPs/issues/165

### CODE REVIEW

File: EIPS/eip-165.md [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+    `true` when `interfaceID` is 0x01ffc9a7 (EIP165 interface)
+    `false` when `interfaceID` is 0xffffffff
+    `true` for any `interfaceID` that this contract implements.
+    `false` for any other `interfaceID`
+
+##### Procedure to determine if a contract implements EIP165 standard.
+
+1. The source contact makes a `CALL` to the destination address with input data: 0x01ffc9a701ffc9a7 value: 0 and gas 30000
+2. If the call fails or return false, the destination contract does not implement EIP165
+3. If the call returns true, a second call is made with input data: 0x01ffc9a7ffffffff
```
[the remark this answers is not in the document]
Comment (jbaylina) [reply EIPS/eip-165.md]: A typical example could be to ask if a wallet implements `ITokenFallback`.  In most cases, this call will `throw` or return false. The `return true` case, I see it very unprovable now, but after metropolis I see the possibility that some day solidity could be able to return values in the fallback function. (This could be very useful for implementing proxy contracts for example). I also do not control other languages like serpent or direct assembly contracts. So because of this possibility and to be sure there is no edge cases, I implemented the "Double call", But if you don't see it probable, we can remove this double check and go straight to point 3 checking for return false and exception.

File: EIPS/eip-165.md [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+
+The contract on address 0x_________ can also be used to determine if a contract implements EIP165, an interface, or a set of interfaces. This contract will use the previous procedure to determine the implementation of those interfaces and will hold a cache to optimize the gas costs of those queries.
+
+This contract will implement the next functions:
+
+    function eip165Supported(address _contract) constant returns (bool);
+
+Returns true if `_contract` supports EIP165
+
+    function interfaceSupported(address _contract, bytes4 _interfaceId) constant returns (bool);
```
[the remark this answers is not in the document]
Comment (jbaylina) [reply EIPS/eip-165.md]: If the contract `throws` (because it is an old one) it will consume all the gas.... So in this case I suspect that it's much more efficient to have a `cache`.  For a new contract (and if we remove the double call), probably it will not be very different.

**Comment by jbaylina** [comment]
Please check https://github.com/ethereum/EIPs/issues/672 as alternative to this EIP.

File: EIPS/eip-165.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (Arachnid) [reply EIPS/eip-165.md]: Personally I don't think it's necessary.

File: EIPS/eip-165.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (Arachnid) [reply EIPS/eip-165.md]: Good point on the `throw`.

**Comment by Arachnid** [comment]
I'd still like to standardise this. ENS can't use #672 without massive ouroboros confusion. ;)

File: EIPS/eip-165.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (jbaylina) [reply EIPS/eip-165.md]: If we assume that we will use a cache, then may be it's not so important to remove the "double call" in terms of performance (It will be done only once by the cache contract). And this way we don't risk for edge cases.

[the remark this answers is not in the document]
Comment (Arachnid) [reply EIPS/eip-165.md]: True; how about rewording it such that it's optional? In cases where you already know the target contract implements `supportsInterface`, there's no point in calling it more than once.

File: EIPS/eip-165.md [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+
+This function must return:
+* `true` when `interfaceID` is 0x01ffc9a7 (EIP165 interface)
+* `false` when `interfaceID` is 0xffffffff
+* `true` for any `interfaceID` that this contract implements.
+* `false` for any other `interfaceID`
+
+##### Procedure to determine if a contract implements EIP165 standard.
+
+1. The source contact makes a `CALL` to the destination address with input data: 0x01ffc9a701ffc9a7 value: 0 and gas 30000. `0x01ffc9a701ffc9a7` is the data for `supportsInterface(0x01ffc9a7)`. Asks if self implements itself.
```
Comment (Arachnid) [inline EIPS/eip-165.md]: I still think this needs to be expressed as an ABI, not as raw bytes.

File: EIPS/eip-165.md [hunk 4]
```diff
@@ -0,0 +1,229 @@
+## Preamble
+
+    EIP: <to be assigned>
```
Comment (Arachnid) [inline EIPS/eip-165.md]: This needs setting.

File: EIPS/eip-165.md (hunk 3 above)
Comment (jbaylina) [reply EIPS/eip-165.md]: The problem is that in solidity you cannot catch the throw inside a call by just calling in the ABI way. You will have to use the `call` by hand and check the result.  I added the explanation in the same line, but I think is worthy to just explain how the internal call is make.   Please fill free to rewrite this paragraph in a way you fill more confortable.

**Comment by fulldecent** [comment]
This function should be `pure` if caching will work!

**Comment by fulldecent** [comment]
^^ @RuFeDoRu @Arachnid

**Comment by fulldecent** [comment]
I don't like how the contract needs to know about the cache. This is surely more complicated than need be and the control flow is inverted.

Just make things simple!

**Comment by fulldecent** [comment]
The specification need not be any more complicated than this:

```
pragma solidity ^0.4.19;

/// @title Interface for contracts conforming to ERC-165: Standard Interface Detection
/// @author XXX, William Entriken (https://phor.net)
/// @dev Draft specification at https://github.com/ethereum/eips/issues/165
/// @dev Limitations in Solidity prevent us from declaring this as an interface
/* interface */ contract ERC165 {
    /// @dev The way to calculate an interface's ID is documented in the ERC
    ///  But we also provide it here for convenience.
    bytes4 internal constant INTERFACE_SIGNATURE_ERC165 = // 0x01ffc9a7
        bytes4(keccak256('supportsInterface(bytes4)'));

    /// @notice Query a contract to see if it supports a given interface
    /// @dev See ERC-165 documentation on how to calculate `_interfaceID`
    /// @param _interfaceID The selected interface to test
    /// @return true if the interface is supported
    function supportsInterface(bytes4 _interfaceID) external pure returns (bool);
}
```

Everything else is just delaying review and stalling this from becoming an actual standard.

**Comment by fulldecent** [comment]
I reject my comment about `pure` being necessary. We have updated the ERC165Cache discussed here. Updated version at https://github.com/ethereum/EIPs/pull/881. We hope you will endorse this as the canonical pull request for the ERC #165.
